### ISSUE TITLE
Crash at hierarchical_info_loss on office dataset

### ISSUE THREAD

**Issue by smart0eddie** [WenkeHuang/RethinkFL#4]
Hi
At the first line of hierarchical_info_loss 
f_pos = np.array(all_f)[all_global_protos_keys == label.item()][0].to(self.device)

The debugger reports
Exception has occurred: ValueError
setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (10,) + inhomogeneous part.

I think it is due to different classes having different number of clusters 
Partition 0: 3 clusters
Partition 0: 2 clusters
Partition 0: 3 clusters
Partition 0: 3 clusters
Partition 0: 2 clusters
Partition 0: 3 clusters
Partition 0: 3 clusters
Partition 0: 2 clusters
Partition 0: 1 clusters
Partition 0: 2 clusters

May you help me with this issue?
Thanks

**Comment by bzHunter** [comment]
i meet the same problem, Have you solved it?

**Comment by smart0eddie** [comment]
sadly, no
The other three comparison methods works fine

**Comment by bzHunter** [comment]
> sadly, no The other three comparison methods works fine

I solved it! You may get help at https://stackoverflow.com/questions/18665873/filtering-a-list-based-on-a-list-of-booleans.

**Comment by smart0eddie** [comment]
> > sadly, no The other three comparison methods works fine
> 
> I solved it! You may get help at https://stackoverflow.com/questions/18665873/filtering-a-list-based-on-a-list-of-booleans.
May you share how you modify the code?
So you firstly select elements from the list and then convert them to tensor?

thanks

**Comment by smart0eddie** [comment]
This should work
I search for the corresponding prototype index separately
(actually the prototype should be sorted before use)

```
f_idx = np.where(all_global_protos_keys == label.item())[0][0]
f_pos = all_f[f_idx].to(self.device)
f_neg = torch.cat([f for i, f in enumerate(all_f) if i != f_idx]).to(self.device)

mean_f_pos = mean_f[f_idx].to(self.device)
```

**Comment by ice1code** [comment]
> This should work  这应该可以工作 I search for the corresponding prototype index separately我分别搜索对应的原型索引 (actually the prototype should be sorted before use)实际上原型在使用前应该排序
> 
> ```
> f_idx = np.where(all_global_protos_keys == label.item())[0][0]
> f_pos = all_f[f_idx].to(self.device)
> f_neg = torch.cat([f for i, f in enumerate(all_f) if i != f_idx]).to(self.device)
> 
> mean_f_pos = mean_f[f_idx].to(self.device)
> ```
So the code should like this? Can you help me?

![Image](https://github.com/user-attachments/assets/a695e962-489d-42dc-a9ed-1bf2e0bde805)

**Comment by smart0eddie** [comment]
yep

**Comment by ice1code** [comment]
Your answer is very important to me. Thank you sincerely.
